components/concurrency_manager/src/lib.rs

```diff
@@ -23,38 +23,115 @@ use std::{
 use txn_types::{Key, Lock, TimeStamp};
 
 pub use self::{
     key_handle::{KeyHandle, KeyHandleGuard},
     lock_table::LockTable,
 };
 
+lazy_static! {
+    static ref MAX_TS_LIMIT_GAUGE: IntGauge = register_int_gauge!(
+        "tikv_concurrency_manager_max_ts_limit",
+        "Current value of max_ts_limit"
+    )
+    .unwrap();
+    static ref MAX_TS_GAUGE: IntGauge =
+        register_int_gauge!("tikv_concurrency_manager_max_ts", "Current value of max_ts").unwrap();
+}
+
 // Pay attention that the async functions of ConcurrencyManager should not hold
 // the mutex.
 #[derive(Clone)]
 pub struct ConcurrencyManager {
     max_ts: Arc<AtomicU64>,
+    max_ts_limit: Arc<AtomicU64>,
     lock_table: LockTable,
+
+    pub panic_on_invalid_max_ts: bool,
 }
 
 impl ConcurrencyManager {
     pub fn new(latest_ts: TimeStamp) -> Self {
+        Self::new_with_config(latest_ts, true)
+    }
+
+    pub fn new_with_config(latest_ts: TimeStamp, panic_on_invalid_max_ts: bool) -> Self {
         ConcurrencyManager {
             max_ts: Arc::new(AtomicU64::new(latest_ts.into_inner())),
+            max_ts_limit: Arc::new(AtomicU64::new(0)),
             lock_table: LockTable::default(),
+            panic_on_invalid_max_ts,
         }
     }
 
     pub fn max_ts(&self) -> TimeStamp {
         TimeStamp::new(self.max_ts.load(Ordering::SeqCst))
     }
 
+    /// Updates max_ts with a timestamp from PD TSO, bypassing the limit check.
+    /// TS from PD must be a valid, thus it doesn't need to be checked.
+    /// It has no effect if max_ts >= new_ts or new_ts is TimeStamp::max().
+    ///
+    /// # Safety
+    /// This method should only be used with timestamps obtained from PD TSO.
+    pub fn update_max_ts_from_pd(&self, new_ts: TimeStamp) {
+        if new_ts != TimeStamp::max() {
+            let new_ts = new_ts.into_inner();
+            MAX_TS_GAUGE.set(self.max_ts.fetch_max(new_ts, Ordering::SeqCst).max(new_ts) as i64);
+        }
+    }
+
     /// Updates max_ts with the given new_ts. It has no effect if
     /// max_ts >= new_ts or new_ts is TimeStamp::max().
-    pub fn update_max_ts(&self, new_ts: TimeStamp) {
-        if new_ts != TimeStamp::max() {
-            self.max_ts.fetch_max(new_ts.into_inner(), Ordering::SeqCst);
+    ///
+    /// To avoid invalid ts breaking the invariants, the new_ts should be
+    /// less than or equal to the max_ts_limit.
+    ///
+    /// # Returns
+    /// - Ok(()): If the update is successful or has no effect
+    /// - Err(limit): If new_ts is greater than the max_ts_limit, returns the
+    ///   current limit value
+    pub fn update_max_ts(&self, new_ts: TimeStamp) -> Result<(), InvalidMaxTsUpdate> {
+        if new_ts == TimeStamp::max() {
+            return Ok(());
+        }
+        let new_ts = new_ts.into_inner();
+        let limit = self.max_ts_limit.load(Ordering::SeqCst);
+        if limit > 0 && new_ts > limit {
+            if self.panic_on_invalid_max_ts {
+                panic!(
```

Comment:
There would be not enough information for the kv requests if panic here directly, is there a way to print the related request information?

Reply:
Passing context information into the concurrency manager is not feasible. So to print related information will require us to panic in the caller side or upper layers. There are 17 callers currently, we address them individually, we cannot guarantee consistent panic handling for future code changes. What do you think?

Reply:
One approach is to pass a tracker to the `update_max_ts` method(or just get it by `with_tracker` like elsewhere). The tracker contains some execution context information, and if a panic occurs, the related tracker information can be printed for debugging.

Reply:
I explored the implementation but adding the required boilerplate code wouldn't be worth it for this error message. Since most cases only need the command type and timestamps, I opted to pass a string instead, which also makes the PR more suitable for cherry-picking to older versions.

Reply:
After this change, TiKV's availability state introduces an additional influencing factor: fetching timestamps from PD. If there is a network partition between TiKV and PD leader for a period, the `update_max_ts` could be rejected continuously, introducing a potential stability risk.

This is still a challenging trade-off between stability risk and correctness risk, difficult to decide.

BTW, if the current TiKV is isolated from the PD leader for a period but communication between regions remains normal, what is TiKV's existing behavior?

Reply:
We can implement a failsafe: after consecutive TSO fetch failures, the max_ts check will be suspended until the limit is updated.